src/nvim/api/autocmd.c

```diff
@@ -183,24 +187,42 @@ ArrayOf(DictAs(get_autocmds__ret)) nvim_get_autocmds(Dict(get_autocmds) *opts, A
     }
   }
 
-  if (opts->buffer.type == kObjectTypeInteger || opts->buffer.type == kObjectTypeBuffer) {
-    buf_T *buf = find_buffer_by_handle((Buffer)opts->buffer.data.integer, err);
+  ObjectType buf_type;
+  if (has_buf) {
+    buf_type = opts->buf.type;
+  } else {
+    buf_type = opts->buffer.type;
+  }
+  if (buf_type == kObjectTypeInteger || buf_type == kObjectTypeBuffer) {
+    Integer buf_integer;
+    if (has_buf) {
+      buf_integer = opts->buf.data.integer;
+    } else {
+      buf_integer = opts->buffer.data.integer;
```

Comment:
instead of adding all these conditions throughout the code, at the top of this function, just do the same approach as you did in Lua, except in C. I.e.,

- If `opts->buf` is provided then ignore `opts->buffer`. 
- If `opts->buffer` is provided (but not `buf`), copy it to `buf`.